Briefcase should work behind a proxy network

daveycrockett opened the pull request:
Fixes #86 

See previous pull request (https://github.com/opendatakit/briefcase/pull/105) for conversation to-date, and how to test.

I made the proxy settings auto-save...if either port or host field loses focus, the settings will auto-update.  I also made the text field for port only allow numbers, so there's no longer a need for a validation pop-up.  Wasn't sure how to handle an empty host value, suggestions welcome.  For now it'll just error out when it attempts to use a hostless proxy, which isn't great, but if someone's entering in proxy settings you have to hope they know what they're doing well enough to put a host name in?

Thanks to @rclakmal for implementing the vast majority of this fix!

lognaturel commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -112,6 +114,16 @@ public static String getBriefcaseDirectoryProperty() {
+
+  public static void setBriefcaseProxyProperty(HttpHost value) {
+    if (value == null) {
+	  Preference.APPLICATION_SCOPED.remove(BRIEFCASE_PROXY_HOST_PROPERTY);
```
Should be spaces, not tabs. I think there are a few more places where tabs are introduced. Would be great to do a pass replacing them!

yanokwa commented on src/org/opendatakit/briefcase/ui/MessageStrings.java:
```diff
@@ -73,4 +73,15 @@
+  public static final String PROXY_SETTINGS_STRING =
+		  "<HTML><U>Proxy Settings</U></HTML>";
+  public static final String PROXY_SET_ADVICE =
+		  ". If you are behind a proxy, try setting up your proxy details through 'Settings' tab.";
```
This will result in two periods if you enter the wrong user name and password. It'll say "Please verify that the URL, your user credentials and your permissions are all correct.."

yanokwa commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -143,4 +155,15 @@ public Preferences preferenceFactory(Class<?> node) {
+  public static HttpHost getBriefCaseProxyConnection() {
+	  String host = Preference.APPLICATION_SCOPED.get(
+			  BriefcasePreferences.BRIEFCASE_PROXY_HOST_PROPERTY,null);
+	  if(host != null){
```
Add an extra space after if

yanokwa commented on src/org/opendatakit/briefcase/ui/SettingsPanel.java:
```diff
@@ -88,6 +171,46 @@ public void actionPerformed(ActionEvent e) {
+        			txtHost.setEnabled(true);
+        			txtPort.setEnabled(true);
+        			if (!StringUtils.isNotEmptyNotNull(txtPort.getText())) {
+        				txtPort.setText("80");
```
Default HTTP proxy port is 8080.

yanokwa commented:
@daveycrockett Thanks for taking on this issue. I really like the elegance of the auto save on focus. Nice and clean!

I can confirm that the basic case of connecting to the proxy works. The things I'd like to see improved as far as the UX.
* For an empty host, what would you think about defaulting to `127.0.0.1`?
* I think we should remove the General Settings and Proxy Settings labels. I don't think they add much.
* I did a quick look into Java UI standards (and what IntelliJ/Android Studio do with proxies), and I think we should re-jigger this layout. Namely, the label should come to the right of the checkbox, and if we are doing that, it makes sense to indent the host name and port number. And then make the text boxes expand relative to the size of the window. Attached is a quick mockup. 
![proxy](https://cloud.githubusercontent.com/assets/32369/26426148/0549fce8-409d-11e7-88da-c4693c67c47e.png)
* Maybe `Use HTTP Proxy` is a better title and if we can do one of these fancy number only widgets, that'd be lovely.
   <img width="175" alt="portnumber" src="https://cloud.githubusercontent.com/assets/32369/26426204/3e947280-409d-11e7-9165-52df6e9d5ce2.png">

daveycrockett commented:
@lognaturel it looks as if the HttpClient object in ServerConnectionInfo is instantiated/set once when I connect to an aggregate server, meaning that if I change settings in between connection and pulling individual form submissions, the changes to proxy settings wouldn't be reflected.  Easiest (read: laziest) way to fix this would be to create a new HttpClient every time we interact with the server, i.e. ServerConnectionInfo.getHttpClient() always calls WebUtils.createHttpClient() with each invocation.  Only other way I can think to do this would be to have settings changes trigger unsetting httpClient in any ServerConnectionInfo objects that are lying around (perhaps promoting httpClient to a static member of the class, or something).  Thoughts?  Without a major refactor I think the lazy option may be the only feasible thing, but I defer to the expertise of others more familiar with briefcase.

daveycrockett commented:
Went ahead and implemented the lazy initializer for httpClient, and incorporated all the other feedback.  Thanks @yanokwa and @lognaturel!  Agree that a lot of the files need a formatting pass to get tabs/spaces consistent, I opted to keep that separate so that the diff is hopefully readable.  If/when this is merged, I'd opt to do a second PR just for formatting.
